sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessEvaluationContext.java

```diff
@@ -357,6 +357,37 @@ public CounterSet getCounters() {
   }
 
   /**
+   * Returns true if the step will not produce additional output.
+   *
+   * <p>If the provided transform produces only {@link IsBounded#BOUNDED}
+   * {@link PCollection PCollections}, returns true if the watermark is at
+   * {@link BoundedWindow#TIMESTAMP_MAX_VALUE positive infinity}.
+   *
+   * <p>If the provided transform produces any {@link IsBounded#UNBOUNDED}
+   * {@link PCollection PCollections}, returns the value of
+   * {@link InProcessPipelineOptions#isShutdownUnboundedProducersWithMaxWatermark()}.
+   */
+  public boolean isDone(AppliedPTransform<?, ?, ?> transform) {
+    boolean done = true;
+    for (PValue output : transform.getOutput().expand()) {
+      if (output instanceof PCollection) {
+        IsBounded bounded = ((PCollection<?>) output).isBounded();
+        if (bounded.equals(IsBounded.BOUNDED)
+            || options.isShutdownUnboundedProducersWithMaxWatermark()) {
+          done &=
```

Comment:
The use of `&=` is a bit weird. As far as I can tell, we aren't relying on all of these iterations being executed. Any reason we don't return earlier? Eg:

```
boolean doneAtMaxTimestamp = bounded.equals(IsBounded.BOUNDED)
    || options.isShutdownUnboundedProducersWithMaxWatermark();
boolean atMaxTimestamp = TIMESTAMP_MAX_VALUE.equals(
    watermarkManager.getWatermarks(transform).getOutputWatermark());
if (doneAtMaxTimestamp && atMaxTimestamp) {
  return false;
}
```

You might also consider putting having an `isDone(PCollection)` that is used by this.